plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -132,6 +130,323 @@ nonZeroArg :: (Integer -> Integer -> Integer) -> Integer -> Integer -> Evaluatio
+touching any costing stuff.
+
+1. The simplest example of an accepted denotation is a monomorphic function that takes values of
+built-in types and returns a value of a built-in type as well. For example
+
+    encodeUtf8 :: Text -> BS.ByteString
+
+You can feed 'encodeUtf8' directly to 'makeBuiltinMeaning' without specifying any types:
+
+    toBuiltinMeaning EncodeUtf8 =
+        makeBuiltinMeaning
+            encodeUtf8
+            <costingFunction>
+
+This will add the builtin, the only two things that remain are implementing costing for this
+builtin (out of the scope of this Note) and handling it within the @Flat DefaultFun@ instance
+(see Note [Stable encoding of PLC]).
+
+2. If the type of the denotation has any constrained type variables in it, all of them need to be
+be instantiated. For example feeding @(+)@ directly to 'makeBuiltinMeaning' will give you an error
+message asking to instantiate constrained type variables, which you can do via an explicit type
+annotation or type application or using any other way of specifying types.
+
+Here's how it looks with a type application instantiating the type variable of @(+)@:
+
+    toBuiltinMeaning AddInteger =
+        makeBuiltinMeaning
+            ((+) @Integer)
+            <costingFunction>
+
+Or we can specify the whole type of the denotation by type-applying 'makeBuiltinMeaning':
+
+    toBuiltinMeaning AddInteger =
+        makeBuiltinMeaning
+            @(Integer -> Integer -> Integer)
+            (+)
+            <costingFunction>
+
+Or we can simply annotate @(+)@ with its monomorphized type:
+    toBuiltinMeaning AddInteger =
+        makeBuiltinMeaning
+            ((+) :: Integer -> Integer -> Integer)
+            <costingFunction>
+
+All of these are equivalent.
+
+3. Unconstrained type variables are fine, you don't need to instantiate them (but you may want to if
+you want some builtin to be less general than what Haskell infers for its denotation). For example
+
+    toBuiltinMeaning IfThenElse =
+        makeBuiltinMeaning
+            (\b x y -> if b then x else y)
+            <costingFunction>
+
+works alright. The inferred Haskell type of the denotation is
+
+    forall a. Bool -> a -> a -> a
+
+whose counterpart in Plutus is
+
+    all a. bool -> a -> a -> a
+
+and unsurprisingly it's the exact Plutus type of the added builtin.
+
+It may seem like getting the latter from the former is entirely trivial, however
+'makeBuiltinMeaning' jumps through quite a few hoops to achieve that and below we'll consider those
+of them that are important to know to be able to use 'makeBuiltinMeaning' in cases that are more
+complicated than a simple monomorphic or polymorphic function. But for now let's talk about a few
+more simple cases.
+
+4. Certain types are not built-in, but can be represented via built-in ones. For example, we don't
+have 'Int' as a built-in, but we have 'Integer' and we can represent the former in terms of the
```

Comment:
> I get that the builtin application machinery handles this case nicely so people can't make mistakes, it just feels like it's the wrong place to put that logic...

What do you think would be the right place?

I can change the existing `KnownTypeIn` instance for `Int` to give the user a `TypeError` specifying how exactly they're supposed to handle `Int` and then the user will follow the instructions and copy-paste some code within the definition of the builtin, but... why? This is more error-prone ("ok, I'm getting some weird error when I'm trying to use `Int`, so I'll just use `fromIntegral`") and quite less convenient. I don't think there's any ironclad reason why we handle `Int` automatically currently, but I don't see at the moment how to change that without making things substantially worse.

There's of course the problem that currently we only special-case `Int` while there's also `Word64` and plenty of other similar types, but I'm having hopes of creating a general `TypeError` saying "this type is neither built-in nor one of the special types, you either want to add a new built-in type or use an existing one, for that do blah-blah-blah".

Reply:
Because it's weird magic that can confuse people. I always get confused by this because I think "wait, we only have `Integer` in our universe, how can this work?" I wouldn't mind having to do it explicitly...

Reply:
Anyway, this is off-topic, I guess.